From 330a7f2bd411d014141980f2a3ebaaebf2b6d576 Mon Sep 17 00:00:00 2001
From: Dai Jingtao <daijingtao1503@phytium.com.cn>
Date: Fri, 21 Jun 2024 16:17:17 +0800
Subject: [PATCH 147/150] drm/phytium: Add function phytium_fbdev_destroy of
 Display Engine support

The purpose of this patch is to add a new function `phytium_fbdev_destroy`
to release resources related to `fbdev_phytium_gem` and modify the
`phytium_fbdev_ops` structure to point to this function through the
`fb_destroy` field while retaining other existing functionalities specified
by `DRM_FB_HELPER_DEFAULT_OPS` and `FB_DEFAULT_IOMEM_OPS`.
This helps to enhance and optimize memory management and
destruction operations involving the PhyTium GPU driver.

Signed-off-by: Dai Jingtao <daijingtao1503@phytium.com.cn>
Signed-off-by: Yang Xun <yangxun@phytium.com.cn>
Signed-off-by: Liu Tao <liutao@phytium.com.cn>
Signed-off-by: Wang Yinfeng <wangyinfeng@phytium.com.cn>
Change-Id: Id174aec340304fffadfe631760fdcf75a60af81c
Signed-off-by: Andrew Powers-Holmes <aholmes@omnom.net>
---
 drivers/gpu/drm/phytium/phytium_fbdev.c | 14 +++++++++++---
 1 file changed, 11 insertions(+), 3 deletions(-)

diff --git a/drivers/gpu/drm/phytium/phytium_fbdev.c b/drivers/gpu/drm/phytium/phytium_fbdev.c
index 3de1c4f34dd4..ce5ead2e6a22 100644
--- a/drivers/gpu/drm/phytium/phytium_fbdev.c
+++ b/drivers/gpu/drm/phytium/phytium_fbdev.c
@@ -18,6 +18,14 @@
 #define	PHYTIUM_MAX_CONNECTOR	1
 #define	helper_to_drm_private(x) container_of(x, struct phytium_display_private, fbdev_helper)
 
+static void phytium_fbdev_destroy(struct fb_info *info)
+{
+	struct drm_fb_helper *helper = info->par;
+	struct phytium_display_private *priv = helper_to_drm_private(helper);
+
+	phytium_gem_free_object(&priv->fbdev_phytium_gem->base);
+}
+
 static int phytium_fbdev_mmap(struct fb_info *info, struct vm_area_struct *vma)
 {
 	struct drm_fb_helper *helper = info->par;
@@ -28,10 +36,10 @@ static int phytium_fbdev_mmap(struct fb_info *info, struct vm_area_struct *vma)
 
 static const struct fb_ops phytium_fbdev_ops = {
 	.owner = THIS_MODULE,
-	DRM_FB_HELPER_DEFAULT_OPS,
 	.fb_mmap = phytium_fbdev_mmap,
-	__FB_DEFAULT_IOMEM_OPS_DRAW,
-
+	.fb_destroy = phytium_fbdev_destroy,
+	 DRM_FB_HELPER_DEFAULT_OPS,
+	 FB_DEFAULT_IOMEM_OPS,
 };
 
 static int
-- 
2.47.0

